### ISSUE TITLE
False positive on Lint/UselessConstantScoping with `MyClass = Class.new`

### ISSUE THREAD

**Issue by pdobb** [rubocop/rubocop#13941]
I'm used to defining sub-classes at the bottom of my file. I'm not worried about private scoping, but Lint/UselessConstantScoping is... but only when I use MyClass = Class.new` syntax.

```ruby
class MyBaseClass
  private

  # The reason I am defining `private` scope:
  attr_reader :name

  # No private scope warning produced:
  class MySubClass
  end

  # Private scope warning produced!:
  MySubClass2 = Class.new
end
```

--------

## Expected behavior

No warning should be produced by Lint/UselessAccessModifier when doing e.g.:

```ruby
class MyBaseClass
  private

  attr_reader :name

  MySubClass2 = Class.new
end
```

## Actual behavior

Lint/UselessAccessModifier produces warning for `MySubClass2 = Class.new` syntax:

```ruby
class MyBaseClass
  private

  attr_reader :name

  MySubClass = Class.new
end
```

```
app/models/cop.rb:6:3: W: Lint/UselessConstantScoping: Useless private access modifier for constant scope.
  MySubClass = Class.new
  ^^^^^^^^^^^^^^^^^^^^^^
```

But not for `class MySubclass2; end` syntax

```ruby
class MyBaseClass
  private

  attr_reader :name

  class MySubClass; end
end
```

## Steps to reproduce the problem

Should already be clear, but:

1. Add a `private` scope modifier.
2. Below that, add a method of some sort. e.g. `attr_reader :name`, but any method should work.
3. Below that, add a new class definition using `MySubClass2 = Class.new` syntax.

Note: Number 2 is there to shift the focus from "Useless private access modifier" to "Useless private access modifier for constant scope." -- the source of the false positive.

## RuboCop version

```
$ rubocop -V
1.73.1 (using Parser 3.3.7.1, rubocop-ast 1.38.1, analyzing as Ruby 3.4, running on ruby 3.4.2) [arm64-darwin24]
  - rubocop-minitest 0.37.1
  - rubocop-performance 1.24.0
  - rubocop-rails 2.30.2
  - rubocop-rake 0.7.1
```

**Comment by Earlopain** [comment]
Is this not an argument for the oposite direction? Both are constants, and as such not affected by `private` above.

**Comment by koic** [comment]
Yeah, this is a false negative because this cop belongs to `Lint`, not `Style`. The cop should detect it as an offense since, despite appearing to be a private class, it is actually a public class.

**Comment by pdobb** [comment]
Thanks. I guess it makes sense what you're saying. Both are cases where the class won't become private, so they should be offenses. And consistency is better, either way. But the proposed fix would probably mean I'd just disable this cop.

My code is not attempting to mark the subclasses as private. They are merely being defined at the bottom of the container class. Which happens to be after the `private` scope is started, (sometimes, not always). I guess the alternative would be to reinitiate the `public` scope before defining subclasses, but that seems like a pain since it was never my presumption that the `private` scoping would affect the subclasses.

e.g. this is a common structure for me:

```ruby
class MyClass
  # public methods ...

  private

  # private methods ...

  # MyClass::MySubClass ...
  class MySubClass
    # ...
  end

  # MyClass::MySubClass2 ...
  class MySubClass2
    # ...
  end

  # MyClass::MySubClass3 ... (much more rarely used style)
  MySubClass3 = Class.new(SomeParentClass)
end
```

Either way, thanks! <3

**Comment by pdobb** [comment]
I guess it would make more sense to me if there just wren't any private methods. Like this:

```
class MyClass
  private

  # MyClass::MySubClass ...
  class MySubClass
    # ...
  end

  # MyClass::MySubClass2 ...
  MySubClass2 = Class.new(SomeParentClass)
end
```

Here, it's worth taking away the `private` scoping.

**Comment by mperham** [comment]
I'm seeing what I consider to be false positives too in the Sidekiq codebase. I will frequently say:

```
class
  method A
  method B
  private
  method C
  const # used by private method D <-- Lint/UselessConstantScoping
  method D
end
```

My intent: because it is below the private modifier, it is considered part of the private API and should not be used by other code. If I put it before private, it looks like I'm explicitly making it public. I really don't want to litter the code with `private_constant` declarations; like monkeypatching I prefer to leave open access in case the user needs to do something unexpected.

I guess I can turn off this Cop but I use Standard because it has such a great taste and doesn't need any config. What should I do?

**Comment by bbatsov** [comment]
I'm guessing we can add some configuration option to allow for mixture of methods and constants after the modifier as @pdobb suggested.

**Comment by tejasbubane** [comment]
> I'm guessing we can add some configuration option to allow for mixture of methods and constants after the modifier

Thanks I'll work on it in my PR https://github.com/rubocop/rubocop/pull/13944

**Comment by pdobb** [comment]
Just wanted to add another practical example of what I still think is a false positive. I figure this is the same scenario as my previous examples, but it's important to note that Structs are Data objects are meant to be defined in this way, per:
- https://docs.rubocop.org/rubocop/cops_style.html#stylestructinheritance
- https://docs.rubocop.org/rubocop/cops_style.html#styledatainheritance

```ruby
class MyClass
  # ...

  private

  def my_private_method
    # ...
  end

  # ...

  # MyClass::MyStruct ...
  MyStruct = Struct.new(:attr1, attr2) do
    # ...
  end

  # MyClass::MyData ...
  MyData = Data.define(:attr1, attr2) do
    # ...
  end
end
```

Both constants result in
> Lint/UselessConstantScoping -- Useless `private` modifier for constant scope.
